File: backend/app/modules/todos/schemas.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from .models import TodoStatus


class TodoTagRead(BaseModel):
    id: UUID
    name: str

    model_config = ConfigDict(from_attributes=True)


class TodoItemBase(BaseModel):
    title: str
    description: str | None = None
    notes: str | None = None
    due_date: datetime | None = None
    status: TodoStatus = Field(default=TodoStatus.todo)
    position: int | None = None
    tags: list[str] | None = None


class TodoItemCreate(TodoItemBase):
    title: str


class TodoItemUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    notes: str | None = None
    due_date: datetime | None = None
    status: TodoStatus | None = None
    position: int | None = None
    tags: list[str] | None = None


class TodoItemRead(BaseModel):
    id: UUID
    list_id: UUID
    title: str
    description: str | None
    notes: str | None
    due_date: datetime | None
    status: TodoStatus
    position: int
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    tags: list[TodoTagRead]

    model_config = ConfigDict(from_attributes=True)


class TodoListBase(BaseModel):
    name: str
    description: str | None = None


class TodoListCreate(TodoListBase):
    name: str


class TodoListUpdate(BaseModel):
    name: str | None = None
    description: str | None = None


class TodoListRead(BaseModel):
    id: UUID
    name: str
    description: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TodoListSummary(TodoListRead):
    item_count: int = 0


class TodoListDetail(TodoListRead):
    items: list[TodoItemRead] = Field(default_factory=list)

File: frontend/src/features/todos/api/schemas.ts
import { z } from "zod";

export const todoStatusSchema = z.enum(["todo", "in_progress", "blocked", "done"]);

export type TodoStatus = z.infer<typeof todoStatusSchema>;

export const todoTagSchema = z.object({
  id: z.string().uuid(),
  name: z.string().min(1),
});

export const todoItemSchema = z.object({
  id: z.string().uuid(),
  list_id: z.string().uuid(),
  title: z.string(),
  description: z.string().nullable().optional(),
  notes: z.string().nullable().optional(),
  due_date: z.string().datetime().nullable().optional(),
  status: todoStatusSchema,
  position: z.number().int().nonnegative(),
  completed_at: z.string().datetime().nullable().optional(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
  tags: z.array(todoTagSchema),
});

export type TodoItem = z.infer<typeof todoItemSchema>;

export const todoListReadSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  description: z.string().nullable().optional(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});

export type TodoListRead = z.infer<typeof todoListReadSchema>;

export const todoListSummarySchema = todoListReadSchema.extend({
  item_count: z.number().int().nonnegative(),
});

export type TodoListSummary = z.infer<typeof todoListSummarySchema>;

export const todoListDetailSchema = todoListSummarySchema.extend({
  items: z.array(todoItemSchema),
});

export type TodoListDetail = z.infer<typeof todoListDetailSchema>;

export const todoListsResponseSchema = z.array(todoListSummarySchema);

export type TodoListsResponse = z.infer<typeof todoListsResponseSchema>;
